Analytics Tracking Prototype

shiki opened the pull request:
Closes #5501. 

@astralbodies @joshheald Here's what I got so far for the algorithm for tracking "usage" of analytics. 

I haven't played enough on this to tune the settings. Though so far it seems to be working nicely? Let me know what you think! 

You can adjust the settings here:

https://github.com/woocommerce/woocommerce-ios/blob/a5fd1cde2307f7d57422cb8eaa02281dcf470296/WooCommerce/Classes/ViewRelated/Dashboard/Stats%20v4/StoreStatsUsageTracksEventEmitter.swift#L12-L30

shiki committed:
Add used_analytics event

shiki committed:
Add initial EventEmitter

shiki committed:
Add more unit tests

shiki committed:
Call interacted() in areas where we consider they're using Analytics

shiki committed:
Display a notice if we tracked used_analytics

shiki committed:
Add comments

joshheald commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsUsageTracksEventEmitter.swift:
```diff
@@ -0,0 +1,75 @@
+    /// The maximum number of seconds in between interactions before we will consider the
+    /// merchant to have been idle. If they were idle, the time and interactions counting
+    /// will be reset.
+    private let idleTimeThreshold: TimeInterval = 10
```
I think it would be safe to make this bit longer than the minimum interaction time. Maybe 15 or 20s

shiki replied:
Increased to 20s in c83325ef6eb0cb12b35308cbb40448327fe17ac2 (draft PR https://github.com/woocommerce/woocommerce-ios/pull/5820)

joshheald reviewed (approved):
Feels pretty good to me in use. Possibly increase the idle time threshold. 

I think it would be good to get opinions from a few more people with different test site data.

shiki commented:
Thank you, @joshheald! 

@jaclync would you mind also taking a look at this? IIRC, you have lots of experience with Stats. So it'd be nice to hear your opinion on this too. 🙂

jaclync commented on WooCommerce/WooCommerceTests/ViewRelated/Dashboard/Stats V4/StoreStatsUsageTracksEventEmitterTests.swift:
```diff
@@ -0,0 +1,183 @@
+    }
+
+    func interacted(at date: String) {
+        let date = dateFormatter.date(from: date)!
```
I like how passing a date string makes it more readable!

jaclync commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsUsageTracksEventEmitter.swift:
```diff
@@ -0,0 +1,75 @@
+    /// - Pull-to-refresh
+    /// - Tapping on the bars in the chart
+    /// - Changing the tab
+    /// - Navigating to the My Store tab
```
How about tapping on a product in the top performers section?

shiki replied:
Added in ea5fa60ea597affb0dc0f4fe02cafaeab1debac2 (draft PR #5820)

jaclync reviewed (approved):
The draft looks good to me, code and testing-wise! 👍 

Just wanted to mention the upcoming Home Screen project that includes updating the dashboard UI p91TBi-6b8-p2. I think most of the interactions still apply, maybe we can add a property for the UI version (before or after Home Screen updates) when the time comes.

shiki commented:
Thank you, @jaclync. 

I'm just waiting for @astralbodies to take a look at this before I apply your suggestion.

astralbodies commented:
@shiki The current settings seem fairly decent. We can always ship this and play around with it if we feel it is severely under or over-counting.

shiki committed:
Merge remote-tracking branch 'origin/develop' into issue/5501-prototype-analytics-tracking

shiki committed:
Merge remote-tracking branch 'origin/trunk' into issue/5501-analytics-tracking-finalize

shiki committed:
Merge remote-tracking branch 'origin/trunk' into issue/5501-prototype-analytics-tracking

shiki committed:
Remove notice call

shiki committed:
Increase idle time threshold to 20

shiki committed:
Merge remote-tracking branch 'origin/issue/5501-prototype-analytics-tracking' into issue/5501-analytics-tracking-finalize

shiki committed:
Include tapping on Top Performers

shiki committed:
Replace Singleton with manual injection

shiki committed:
Move My Store tap interaction detection to viewWillAppear

While not the best, I think this is good enough so that we can keep the
UsageTracksEventEmitter dependency within the Stats-related code.

shiki committed:
Add scroll view interaction tracking to Old stats period VC

shiki committed:
Add tab change interaction tracking to old StoreStatsPeriod VC

shiki committed:
Add viewWillAppear interaction tracking to old StatsAndTopPerf VC

shiki committed:
Remove viewWillAppear interaction detection

The StoreStatsV4PeriodViewController.trackChangedTabIfNeeded() is more than
enough to detect if the merchant navigated back to My Store.

shiki committed:
Merge pull request #5820 from woocommerce/issue/5501-analytics-tracking-finalize

Finalize Analytics Usage Tracking

shiki commented:
Applied feedback and finalized the code in #5820. Just waiting for the green build to merge this.

astralbodies commented:
🥳
